File: AssetManager/UserInterface/Forms/GK_Updater/PackFileForm.vb
Imports System.ComponentModel
Imports System.IO
Imports System.Threading
Public Class PackFileForm
    Public Property PackVerified As Boolean = False
    Private Working As Boolean = False
    Private PackFunc As New ManagePackFile
    Sub New(ShowFunctions As Boolean)
        InitializeComponent()
        FunctionPanel.Visible = ShowFunctions
        If Not ShowFunctions Then
            CheckPackFile()
        End If
    End Sub
    Private Async Sub CheckPackFile()
        Try
            Working = True
            PackVerified = Await PackFunc.ProcessPackFile()
            Working = False
            If Me.Modal Then Me.Close()
        Catch ex As Exception
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
        End Try
    End Sub
    Private Sub ProgressTimer_Tick(sender As Object, e As EventArgs) Handles ProgressTimer.Tick
        Try
            If PackFunc.Progress.Percent > 0 Then
                ProgressBar.Value = PackFunc.Progress.Percent
                PackFunc.Progress.Tick()
                If PackFunc.Progress.Throughput > 0 And PackFunc.Progress.Percent < 100 Then
                    If Not SpeedLabel.Visible Then SpeedLabel.Visible = True
                    SpeedLabel.Text = PackFunc.Progress.Throughput.ToString & " MB/s"
                Else
                    SpeedLabel.Visible = False
                End If
            End If
            StatusLabel.Text = PackFunc.Status
        Catch ex As Exception
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
        End Try
    End Sub
    Private Sub VerifyPackButton_Click(sender As Object, e As EventArgs) Handles VerifyPackButton.Click
        Try
            CheckPackFile()
        Catch ex As Exception
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
        End Try
    End Sub
    Private Async Sub NewPackFile()
        Try
            Working = True
            If Not Await PackFunc.CreateNewPackFile() Then
                Message("Error while creating a new pack file.", vbOKOnly + vbExclamation, "Error", Me)
            End If
            Working = False
        Catch ex As Exception
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
        End Try
    End Sub
    Private Async Sub NewPackButton_Click(sender As Object, e As EventArgs) Handles NewPackButton.Click
        Try
            NewPackFile()
        Catch ex As Exception
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
        End Try
    End Sub

    Private Sub PackFileForm_Closing(sender As Object, e As CancelEventArgs) Handles Me.Closing
        e.Cancel = Working
    End Sub
End Class

File: AssetManager/UserInterface/Forms/GK_Updater/PackFileForm.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using System.Linq;
using System.Threading.Tasks;
using System.ComponentModel;
using System.Data.Common;
using System.IO;
using AssetManager.UserInterface.CustomControls;

namespace AssetManager.UserInterface.Forms.GK_Updater
{


    public partial class PackFileForm : ExtendedForm
    {
        public bool PackVerified { get; set; }
        private bool Working = false;

        private ManagePackFile PackFunc = new ManagePackFile();
        public PackFileForm(bool showFunctions)
        {
            InitializeComponent();
            FunctionPanel.Visible = showFunctions;
            if (!showFunctions)
            {
                CheckPackFile();
            }
        }

        private async void CheckPackFile()
        {
            try
            {
                Working = true;
                PackVerified = await PackFunc.ProcessPackFile();
                Working = false;
                if (this.Modal)
                    this.Close();
            }
            catch (Exception ex)
            {
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
            }
        }

        private void ProgressTimer_Tick(object sender, EventArgs e)
        {
            try
            {
                if (PackFunc.Progress.Percent > 0)
                {
                    ProgressBar.Value = PackFunc.Progress.Percent;
                    PackFunc.Progress.Tick();
                    if (PackFunc.Progress.Throughput > 0 & PackFunc.Progress.Percent < 100)
                    {
                        if (!SpeedLabel.Visible)
                            SpeedLabel.Visible = true;
                        SpeedLabel.Text = PackFunc.Progress.Throughput.ToString() + " MB/s";
                    }
                    else
                    {
                        SpeedLabel.Visible = false;
                    }
                }
                StatusLabel.Text = PackFunc.Status;
            }
            catch (Exception ex)
            {
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
            }
        }

        private void VerifyPackButton_Click(object sender, EventArgs e)
        {
            try
            {
                var GKFormInstance = Helpers.ChildFormControl.GKUpdaterInstance();
                if (!GKFormInstance.ActiveUpdates())
                {
                    CheckPackFile();
                }
                else
                {
                    OtherFunctions.Message("This process will interfere with the active running updates. Please stop all updates and try again.", (int)MessageBoxButtons.OK + (int)MessageBoxIcon.Exclamation, "Cannot Continue", this);
                }
            }
            catch (Exception ex)
            {
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
            }
        }

        private async void NewPackFile()
        {
            try
            {
                Working = true;
                if (!await PackFunc.CreateNewPackFile())
                {
                    OtherFunctions.Message("Error while creating a new pack file.", (int)MessageBoxButtons.OK + (int)MessageBoxIcon.Exclamation, "Error", this);
                }
                Working = false;
            }
            catch (Exception ex)
            {
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
            }
        }

        private void NewPackButton_Click(object sender, EventArgs e)
        {
            try
            {
                NewPackFile();
            }
            catch (Exception ex)
            {
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
            }
        }

        private void PackFileForm_FormClosing(object sender, FormClosingEventArgs e)
        {
            e.Cancel = Working;
        }
    }
}